devito/types/array.py

```diff
@@ -204,6 +204,8 @@ class ArrayMapped(Array):
                                           (_C_field_dmap, c_void_p),
                                           (_C_field_size, c_uint64)]}))
 
+    _C_typedata = 'struct ' + _C_structname
```

Comment:
I designed the `_C_*` API so that a class only needs to override `_C_ctype`

You should instead add some logic here: 
https://github.com/devitocodes/devito/blob/master/devito/types/basic.py#L82